src/Giraffe/WebSockets.fs

```diff
@@ -0,0 +1,182 @@
+            let buffer = Array.zeroCreate messageSize |> ArraySegment<byte>
+            use memoryStream = new IO.MemoryStream()
+            let mutable endOfMessage = false
+            let mutable keepRunning = Unchecked.defaultof<_>
+
+            while not endOfMessage do
+                let! received = reference.WebSocket.ReceiveAsync(buffer, cancellationToken)
+                if received.CloseStatus.HasValue then
+                    do! reference.WebSocket.CloseAsync(received.CloseStatus.Value, received.CloseStatusDescription, cancellationToken)
+                    keepRunning <- false
+                    endOfMessage <- true
+                else
+                    memoryStream.Write(buffer.Array,buffer.Offset,received.Count)
+                    if received.EndOfMessage then
+                        match received.MessageType with
+                        | WebSocketMessageType.Binary ->
+                            raise (NotImplementedException())
+                        | WebSocketMessageType.Close ->
+                            keepRunning <- false 
+                            endOfMessage <- true
+                        | WebSocketMessageType.Text ->
+                            let! r = 
+                                memoryStream.ToArray()
+                                |> System.Text.Encoding.UTF8.GetString
+                                |> fun s -> s.TrimEnd(char 0)
+                                |> messageF reference
+
+                            keepRunning <- true
+                            endOfMessage <- true
+                        | _ ->
+                            raise (NotImplementedException())
+
+            return keepRunning
+        }
+
+        member private this.RegisterClient<'Msg>(reference:WebSocketReference,connectedF: WebSocketReference -> Task<unit>,messageF,cancellationToken:CancellationToken) = task {
+            connections.AddOrUpdate(reference.ID, reference, Func<_,_,_>(fun _ _ -> reference)) |> ignore
+            do! connectedF reference
+            let mutable running = true
+            while running && not cancellationToken.IsCancellationRequested do
+                let! msg = this.Receive<'Msg>(reference,messageF,cancellationToken)
+                running <- msg
+
+            match connections.TryRemove reference.ID with
+            | true, reference -> do! reference.CloseAsync()
+            | _ -> ()
+        }
+
+        /// Creates a new WebSocket connection and negotiates the subprotocol.
+        member this.CreateSocket(onConnected,onMessage,?webSocketID,?supportedProtocols:seq<WebSocketSubprotocol>,?cancellationToken) =
+            let negotiateSubProtocol(requestedSubProtocols,supportedProtocols:seq<WebSocketSubprotocol>) =
+                supportedProtocols
+                |> Seq.tryFind (fun (supported:WebSocketSubprotocol) ->
+                    requestedSubProtocols |> Seq.contains supported.Name)
```

Comment:
O(N^2). Is it OK here?

Reply:
is "negotiateSubProtocol" even used, I cannot see where fun it is used in the method scope? I think may have been a helper function but below you manually implement same code L171.

Given this is run on many requests and slow may need to think of a better structure, on 1000s of sockets would hurt perf. This would be a good candidate for creating a scoped Set value from seq (before fun returned so captured in closure)

No big deal but you can probably move "run" L159 out of HttpHandler, above into constructor and pass the 'next' 'ctx' args, so we are not creating that fun every time for no good reason?

Reply:
removed the duplication